web/src/hashchange.js

```diff
@@ -201,6 +201,13 @@ function do_hashchange_normal(from_reload) {
                     narrow_opts.then_select_offset = page_params.initial_narrow_offset;
                 }
             }
+            const location_data_for_hash =
+                browser_history.hash_to_page_location[window.location.hash];
+            if (location_data_for_hash !== undefined) {
+                message_lists.home.pre_narrow_offset = location_data_for_hash.offset;
+                narrow_opts.then_select_id = location_data_for_hash.narrow_pointer;
+                narrow_opts.then_select_offset = location_data_for_hash.narrow_offset;
```

Comment:
Can you explain what the `pre_narrow_offset` logic here is doing? I don't understand what it could be doing -- it seems to me that `then_select_id` plus `then_select_offset` will determine what scroll position we end up at when entering a view, if provided?

Reply:
I took this from the refresh logic:

```js

            if (from_reload) {
                blueslip.debug("We are narrowing as part of a reload.");
                if (page_params.initial_narrow_pointer !== undefined) {
                    message_lists.home.pre_narrow_offset = page_params.initial_offset;
                    narrow_opts.then_select_id = page_params.initial_narrow_pointer;
                    narrow_opts.then_select_offset = page_params.initial_narrow_offset;
                }
            }

```

But yeah it seems to work fine without it! It seems like these weren't meant to be used together, so maybe it can be removed from the reload logic as well?

Reply:
Oh I see -- it's only relevant in views that aren't message narrows, so that's not relevant here.

In that case I think I can do:

```diff

diff --git a/web/src/browser_history.js b/web/src/browser_history.js
index 1c68caf686..7b5f8aafe2 100644
--- a/web/src/browser_history.js
+++ b/web/src/browser_history.js
@@ -103,24 +103,19 @@ export const hash_to_page_location = {};
 
 export function save_narrow() {
     const narrow_data = {};
-    const $row = message_lists.home.selected_row();
     if (!narrow_state.active()) {
-        if ($row.length > 0) {
-            narrow_data.offset = $row.get_offset_to_window().top;
-        }
-    } else {
-        narrow_data.offset = message_lists.home.pre_narrow_offset;
-
-        // narrow_state.active() is true, so this is the current
-        // narrowed message list.
-        const narrow_pointer = message_lists.current.selected_id();
-        if (narrow_pointer !== -1) {
-            narrow_data.narrow_pointer = narrow_pointer;
-        }
-        const $narrow_row = message_lists.current.selected_row();
-        if ($narrow_row.length > 0) {
-            narrow_data.narrow_offset = $narrow_row.get_offset_to_window().top;
-        }
+        return;
+    }
+
+    const narrow_pointer = message_lists.current.selected_id();
+    if (narrow_pointer === -1) {
+        return;
+    }
+    narrow_data.narrow_pointer = narrow_pointer;
+    const $narrow_row = message_lists.current.selected_row();
+    if ($narrow_row.length === 0) {
+        return;
     }
+    narrow_data.narrow_offset = $narrow_row.get_offset_to_window().top;
     hash_to_page_location[window.location.hash] = narrow_data;
 }
```

Reply:
Yeah that looks more plausible. I think once we make narrowing deactivation no longer a different code path, this is definitely the sort of thing we'll want.